commit 900e251fcdaad7f1fcd9ee9158312740ce31b8ec (refs/changes/52/852/3)
Author: Bryan Cantrill <bryan@joyent.com>
Date:   2016-11-15T14:48:47-08:00 (2 years, 11 months ago)
    
    TOOLS-1602 v8plus dies when C constructors generate exceptions
    Reviewed by: Joshua M. Clulow <jmc@joyent.com>
    Approved by: Joshua M. Clulow <jmc@joyent.com>

diff --git a/CHANGES.md b/CHANGES.md
index ffaf79f..e802ed5 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,10 @@
 # v8plus Change History
 
+## 1.0.1
+
+A minor bug fix that allows C constructors to indicate an exception without
+inducing a core dump.
+
 ## 1.0.0
 
 This is the first major release of *v8plus*.  Throughout its development
diff --git a/package.json b/package.json
index 9dbe585..2b92e16 100644
--- a/package.json
+++ b/package.json
@@ -1,10 +1,11 @@
 {
-	"author": "Keith M Wesolowski <keith.wesolowski@joyent.com>",
+	"author": "Keith 'The Mayor' Wesolowski <keith.wesolowski@joyent.com>",
 	"name": "v8plus",
 	"description": "utility environment for writing addons in C",
-	"version": "1.0.0",
+	"version": "1.0.1",
 	"contributors": [
-		"Joshua M. Clulow <jmc@joyent.com>"
+		"Joshua M. Clulow <jmc@joyent.com>",
+		"Bryan Cantrill <bryan@joyent.com>"
 	],
 	"repository": {
 		"type": "git",
diff --git a/v8plus_objectwrap.cc b/v8plus_objectwrap.cc
index eadd345..62cd05a 100644
--- a/v8plus_objectwrap.cc
+++ b/v8plus_objectwrap.cc
@@ -234,6 +234,16 @@ v8plus::ObjectWrap::~ObjectWrap()
 	v8plus_module_defn_t *mdp =
 	    reinterpret_cast<v8plus_module_defn_t *>(_defn);
 
+	if (_c_impl == NULL) {
+		/*
+		 * If the pointer is NULL, it means we are in the ObjectWrap
+		 * destructor after the C constructor indicated an exception.
+		 * In this case, we don't call the C destructor -- and indeed
+		 * _defn is uninitialized so we must not attempt to use it.
+		 */
+		return;
+	}
+
 	mdp->vmd_dtor(_c_impl);
 	(void) _objhash.erase(_c_impl);
 }
